### PR TITLE
[gardenlet] Switch `Shoot` migration controller to `controller-runtime`

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity scalability
/kind enhancement

**What this PR does / why we need it**:
Refactor the `Shoot` migration controller to controller-runtime.
The other two controllers (care, main) are not part of this PR and will follow separately.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:
Generally, we want to follow this cookbook while refactoring existing controllers:

- Add documentation
- Add integration test based on envtest (if not already present)
- Switch controller to controller-runtime

This PR is very similar to parts of https://github.com/gardener/gardener/pull/6926.
/cc @shafeeqes 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by rfranzke** [comment]
/test pull-gardener-e2e-kind

ref #7064

File: pkg/gardenlet/controller/backupentry/add.go [hunk 1]
```diff
@@ -44,7 +44,8 @@ func AddToManager(
 
 	if gardenletfeatures.FeatureGate.Enabled(features.ForceRestore) && confighelper.OwnerChecksEnabledInSeedConfig(cfg.SeedConfig) {
 		if err := (&migration.Reconciler{
-			Config: cfg,
+			Config:   *cfg.Controllers.BackupEntryMigration,
+			SeedName: cfg.SeedConfig.Name,
```
Comment (shafeeqes) [inline pkg/gardenlet/controller/backupentry/add.go:48]: Thanks!

File: docs/concepts/gardenlet.md [hunk 2]
```diff
@@ -359,6 +359,21 @@ If the connection to the `/healthz` endpoint or the update of the `Lease` fails
...
 
+### [`Shoot` Controller](../../pkg/gardenlet/controller/shoot)
+
+The `Shoot` controller in the `gardenlet` reconciles `Shoot` objects with the help of the following reconcilers.
+
+#### "Migration" Reconciler
+
+This reconciler is only active if the [`ForceRestore`](../deployment/feature_gates.md#list-of-feature-gates) feature gate is enabled in the `gardenlet` and if the `Seed` has owner checks enabled (i.e., `spec.setttings.ownerChecks.enabled=true`).
+It checks if the source `Seed` also has owner checks enabled.
+If not or if the `BackupEntry` is being deleted, it sets the `status.migrationStartTime` to `nil`.
```
Comment (shafeeqes) [inline docs/concepts/gardenlet.md]:
```suggestion
It checks if the source `Seed` also has owner checks enabled. If not or if the `Shoot` is being deleted, it sets the `status.migrationStartTime` to `nil`.
```

File: pkg/controllerutils/predicate/predicate_test.go [hunk 3]
```diff
@@ -495,4 +501,95 @@ var _ = Describe("Predicate", func() {
...
+			verifyPredicate(p, obj, gomega.BeFalse())
+		})
+
+		It("should return true if the source seed has owner checks enabled", func() {
+			gomega.Expect(fakeClient.Create(ctx, sourceSeed)).To(gomega.Succeed())
+
+			verifyPredicate(p, obj, gomega.BeTrue())
+		})
+
+		It("should return true if the source seed has owner checks disabled", func() {
```
Comment (shafeeqes) [inline pkg/controllerutils/predicate/predicate_test.go]:
```suggestion
		It("should return false if the source seed has owner checks disabled", func() {
```

File: pkg/controllerutils/predicate/predicate_test.go [hunk 4]
```diff
@@ -495,4 +501,95 @@ var _ = Describe("Predicate", func() {
...
+			}
+		})
+
+		It("should return false if status.SeedName is nil", func() {
+			obj.Status.SeedName = nil
+
+			verifyPredicate(p, obj, gomega.BeFalse())
+		})
+
+		It("should return false if spec.Status.SeedName status.SeedName are equal", func() {
```
Comment (shafeeqes) [inline pkg/controllerutils/predicate/predicate_test.go]:
```suggestion
		It("should return false if spec.SeedName and status.SeedName are equal", func() {
```

File: pkg/utils/gardener/migration_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+			}
+		})
+
+		It("should return false if status.SeedName is nil", func() {
+			obj.Status.SeedName = nil
+
+			Expect(IsObjectBeingMigrated(ctx, fakeClient, obj, seedName, getSeedNamesFromObject)).To(BeFalse())
+		})
+
+		It("should return false if spec.Status.SeedName status.SeedName are equal", func() {
```
Comment (shafeeqes) [inline pkg/utils/gardener/migration_test.go]:
```suggestion
		It("should return false if spec.SeedName and status.SeedName are equal", func() {
```

File: pkg/utils/gardener/migration_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			Expect(IsObjectBeingMigrated(ctx, fakeClient, obj, seedName, getSeedNamesFromObject)).To(BeFalse())
+		})
+
+		It("should return true if the source seed has owner checks enabled", func() {
+			Expect(fakeClient.Create(ctx, sourceSeed)).To(Succeed())
+
+			Expect(IsObjectBeingMigrated(ctx, fakeClient, obj, seedName, getSeedNamesFromObject)).To(BeTrue())
+		})
+
+		It("should return true if the source seed has owner checks disabled", func() {
```
Comment (shafeeqes) [inline pkg/utils/gardener/migration_test.go]:
```suggestion
		It("should return false if the source seed has owner checks disabled", func() {
```

File: pkg/utils/gardener/migration_test.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			sourceSeed.Spec.Settings.OwnerChecks.Enabled = false
+			Expect(fakeClient.Create(ctx, sourceSeed)).To(Succeed())
+
+			Expect(IsObjectBeingMigrated(ctx, fakeClient, obj, seedName, getSeedNamesFromObject)).To(BeFalse())
+		})
```
Comment (shafeeqes) [inline pkg/utils/gardener/migration_test.go]: Maybe we can reduce the duplication and keep the unit tests only here.

File: pkg/gardenlet/controller/shoot/migration/reconciler.go [hunk 8]
```diff
@@ -12,85 +12,41 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package shoot
+package migration
```
Comment (shafeeqes) [inline pkg/gardenlet/controller/shoot/migration/reconciler.go:15]: 2022 SAP SE

**Review by shafeeqes — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/controller/shoot/migration/add.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	// It's not possible to overwrite the event handler when using the controller builder. Hence, we have to build up
+	// the controller manually.
+	c, err := controller.New(
+		ControllerName,
+		mgr,
+		controller.Options{
+			Reconciler:              r,
+			MaxConcurrentReconciles: 1,
+			RecoverPanic:            true,
+			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(time.Millisecond, 2*time.Second),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/migration/add.go]: @shafeeqes Why do we need this (I copied this from the `BackupEntry` migration controller)?

Comment (shafeeqes) [reply pkg/gardenlet/controller/shoot/migration/add.go]: I cannot see it in the master ? :eyes: https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/backupentry/migration/add.go 
It's present in https://github.com/gardener/gardener/blob/98ce217fd6bc916e9f08fb0d63d6480ab13ea7ab/pkg/gardenlet/controller/seed/lease/add.go#L52 though.

Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/migration/add.go]: Interesting 😄 It's gone now, PTAL

**Review by shafeeqes — COMMENTED** [review]
/lgtm

**Pull request merged**